presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/ElasticsearchQueryBuilder.java

```diff
@@ -39,37 +43,43 @@
 import static io.prestosql.spi.type.BooleanType.BOOLEAN;
 import static io.prestosql.spi.type.DoubleType.DOUBLE;
 import static io.prestosql.spi.type.IntegerType.INTEGER;
+import static io.prestosql.spi.type.RealType.REAL;
+import static io.prestosql.spi.type.SmallintType.SMALLINT;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP;
+import static io.prestosql.spi.type.TinyintType.TINYINT;
 import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static java.time.format.DateTimeFormatter.ISO_DATE_TIME;
 
 public class ElasticsearchQueryBuilder
 {
-    private ElasticsearchQueryBuilder() {}
+    private ElasticsearchQueryBuilder()
+    {
+    }
 
-    public static QueryBuilder buildSearchQuery(TupleDomain<ColumnHandle> constraint, List<ElasticsearchColumnHandle> columns, Optional<String> query)
+    public static QueryBuilder buildSearchQuery(ConnectorSession session, TupleDomain<ColumnHandle> constraint, Optional<String> query)
     {
-        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
-        for (ElasticsearchColumnHandle column : columns) {
-            BoolQueryBuilder columnQueryBuilder = new BoolQueryBuilder();
-            Type type = column.getType();
-            if (constraint.getDomains().isPresent()) {
-                Domain domain = constraint.getDomains().get().get(column);
-                if (domain != null) {
-                    columnQueryBuilder.should(buildPredicate(column.getName(), domain, type));
+        BoolQueryBuilder queryBuilder = new BoolQueryBuilder();
+        if (constraint.getDomains().isPresent()) {
+            for (Map.Entry<ColumnHandle, Domain> entry : constraint.getDomains().get().entrySet()) {
+                ElasticsearchColumnHandle column = (ElasticsearchColumnHandle) entry.getKey();
```

Comment:
You could have `constraint.transform(ElasticsearchColumnHandle.class::cast)` at the caller side to avoid casting here. Ideally, all the "ColumnHandle" -> "ElasticsearchColumnHandle" should be handled as early as possible.